matrix_sdk/src/event_handler.rs

```diff
@@ -0,0 +1,397 @@
+// Copyright 2020 Damir Jelić
+// Copyright 2020 The Matrix.org Foundation C.I.C.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Types and traits related for event handlers. For usage, see
+//! [`Client::register_event_handler`].
```

Comment:
Can we get some high level overview/docs that describe how this works, the new handler is a bit of type magic. While I do get the gist of it a description wouldn't hurt, the doc doesn't need to be necessarily public though I don't think anyone would mind them publicly either.

While we're here, it would be nice to get some tests for the handler in this module, it's a bit weird that the only tests the handler now has reside in the app service crate.

Reply:
I'll try to improve docs with some details about the type magic.

Reply:
Wdyt? https://github.com/matrix-org/matrix-rust-sdk/pull/309/commits/606a57203ea280d0284584bba8d72752e0ab8dcd

Reply:
Looks good, thanks.